src/main/java/algorithms/curated170/easy/ReadNCharactersGivenRead4.java

```diff
@@ -1,36 +1,32 @@
 package algorithms.curated170.easy;
-class Reader4{
-      int read4(char[] buf){
-          
-          //TODO
-          return -1;
-      }
-} 
+
 public class ReadNCharactersGivenRead4 extends Reader4 {
-    /**
-     * @param buf Destination buffer
-     * @param n   Number of characters to read
-     * @return    The number of actual characters read
-     */
-public int read(char[] buf, int n) {
-  boolean eof = false;      // end of file flag
-  int total = 0;            // total bytes have read
-  char[] tmp = new char[4]; // temp buffer
-  
-  while (!eof && total < n) {
-    int count = read4(tmp);
-    
-    // check if it's the end of the file
-    eof = count < 4;
-    
-    // get the actual count
-    count = Math.min(count, n - total);
-    
-    // copy from temp buffer to buf
-    for (int i = 0; i < count; i++) 
-      buf[total++] = tmp[i];
-  }
-  
-  return total;
-}
+    public int read(char[] buf, int n) { 
+        int index = 0;
+        int length = 0;
+        char[] buf4 = new char[4];
+        int copiedChars = 0;  
+       
+        while (copiedChars < n) {
+       
+            if (index == length) {
+                length = read4(buf4);
+                index = 0;
+                if (length == 0) {
+                    break;
+                }
+            }
+            
+            buf[copiedChars++] = buf4[index++];
```

Comment:
read4 should read 4 chars at a time but we are only incrementing by 1 in our destination char array.  Why don't we just call it then loop through the 4 or less elements of buf4 to assign them to buf, instead of creating a gap between length and index?

Reply:
Similar to the last solution to some degree.

Reply:
Could you share the code in your mind?

Reply:
```java
    public int read(char[] buf, int n) {
        int copiedChars = 0;
        char[] buf4 = new char[4];
        while(copiedChars< n){
            int length = read4(buf4);
            for(int i=0; i<length && copiedChars < n; i++){
                buf[copiedChars++] = r4[i];
            }
            if(length < 4) break;
        }

        return copiedChars;
    }
```

Reply:
Oke. This is also added as the second approach.